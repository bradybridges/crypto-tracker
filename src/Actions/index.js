export const updateCryptos = (cryptos) => ({
  type: 'UPDATE_CRYPTOS',
  cryptos,
});

export const updateError = (error) => ({
  type: 'UPDATE_ERROR',
  error,
});

export const updateTimeFrame = (timeFrame) => ({
  type: 'UPDATE_TIME_FRAME',
  timeFrame,
});

export const updateTrackedCoins = (symbols) => ({
  type: 'UPDATE_TRACKED_COINS',
  symbols,
});

export const updatePortfolio = (portfolio) => ({
  type: 'UPDATE_PORTFOLIO',
  portfolio,
});
