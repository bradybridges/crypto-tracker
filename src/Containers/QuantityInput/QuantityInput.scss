.qty-input {
  width: 50%;
  height: 60px;
}

.coin-input-div {
  width: 90%;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.coin-label {
  font-size: 1.75em;
}