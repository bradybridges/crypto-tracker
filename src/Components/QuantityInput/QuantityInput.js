import React, { Component } from 'react';
import { connect } from 'react-redux';
import { updatePortfolio } from '../../Actions/index';
import './QuantityInput.scss';

export class QuantityInput extends Component {
  constructor() {
    super();
    
    this.state = {
      name: '',
      qty: '',
    }
  }

    handleChange = (e) => {
    const qty = Number(e.target.value);
    const name = this.props.coin.name;
    const indexExistingCoin = this.props.portfolio.findIndex( coin => coin.name === name);

    if(qty > 0){
      this.setState({name, qty})
      this.updatePortfolio(name, qty, indexExistingCoin);
    } else if( qty === 0) {
      this.setState({name: '', qty: ''});
    }
  }

  updatePortfolio = (name, qty, indexExistingCoin) => {
    if(indexExistingCoin !== -1) {
      let portfolio = this.props.portfolio.map( coin => coin);
      portfolio[indexExistingCoin] = {name, qty};
      this.props.updatePortfolio(portfolio);
      localStorage.setItem('portfolio', JSON.stringify(portfolio));
    } else {
      this.props.updatePortfolio([...this.props.portfolio, {name, qty}]);
      localStorage.setItem('portfolio', JSON.stringify([...this.props.portfolio, {name, qty}]))
    }
  }

  render() {
    const { coin } = this.props;
    return (
      <div className='coin-input-div'>
        <label>{coin.name}</label>
        <input type='number' placeholder={`Quantity: (${this.props.qty})`} onChange={this.handleChange} value={this.state.qty} />
      </div>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  updatePortfolio: portfolio => dispatch( updatePortfolio(portfolio) ),
});

export const mapStateToProps = state => ({
  portfolio: state.portfolio,
});

export default connect(mapStateToProps, mapDispatchToProps)(QuantityInput);