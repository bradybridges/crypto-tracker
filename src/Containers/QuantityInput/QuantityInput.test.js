import React from 'react';
import { shallow } from 'enzyme';
import { QuantityInput, mapDispatchToProps, mapStateToProps } from './QuantityInput';
import { updatePortfolio } from '../../Actions/index';

describe('QuantityInput', () => {
  let wrapper;
  let mockCoin;
  const mockUpdatePortfolio = jest.fn();
  const mockPortfolio = [{name: 'Bitcoin', qty: 1}, {name: 'Litecoin', qty: 0}]
  beforeEach(() => {
    mockCoin = {
      name: 'Bitcoin',
    };
    wrapper = shallow(<QuantityInput updatePortfolio={mockUpdatePortfolio} portfolio={mockPortfolio} coin={mockCoin} qty={5} />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have state default properties of name and qty', () => {
    expect(wrapper.state('name')).toEqual('');
    expect(wrapper.state('qty')).toEqual('');
  });

  it('handleChange should updateState and call updatePortfolio, should update portfolio if coin exists', () =>{ 
     wrapper.instance().updatePortfolio = jest.fn();
     const mockEvent = {
       target: { value: '10' }
     };
     wrapper.instance().handleChange(mockEvent);
     expect(wrapper.state('name')).toEqual('Bitcoin');
     expect(wrapper.state('qty')).toEqual(10);
     expect(wrapper.instance().updatePortfolio).toHaveBeenCalledWith('Bitcoin', 10, 0);
  });

  it('handleChange should not call updatePortfolio if qty is 0', () => {
    wrapper.instance().updatePortfolio = jest.fn();
    const mockEvent = {
      target: {value: '0'},
    };
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('name')).toEqual('');
    expect(wrapper.state('qty')).toEqual('');
    expect(wrapper.instance().updatePortfolio).not.toHaveBeenCalled();
  });

  it('updatePortfolio should update coin if it already exists', () => {
    const expectedPortfolio = [{name: "Bitcoin", qty: 20}, {name: "Litecoin", qty: 0}];
    wrapper.instance().updatePortfolio('Bitcoin', 20, 0);
    expect(wrapper.instance().props.updatePortfolio).toHaveBeenCalledWith(expectedPortfolio);
  });

  it('updatePortfolio should add new coin if it does not exist', () =>{
    const expectedPortfolio = [{name: "Bitcoin", qty: 1}, {name: "Litecoin", qty: 0}, {name: 'Ripple', qty: 3}];
    wrapper.instance().updatePortfolio('Ripple', 3, -1);
    expect(wrapper.instance().props.updatePortfolio).toHaveBeenCalledWith(expectedPortfolio);
  });

  it.skip('when quantity input changes handleChange should be called', () => {
    const mockEvent = {
      target: {value: '10'}
    };
    wrapper.instance().handleChange = jest.fn();
    wrapper.find('.qty-input').simulate('change', mockEvent);
    expect(wrapper.instance().handleChange).toHaveBeenCalled();
  });
});

describe('mapDispatchToProps', () => {
  it('should map updatePortfolio to props', () => {
    const mockDispatch = jest.fn();
    const mockPortfolio = [
      {name: 'Bitcoin', qty: 1},
      {name: 'Litecoin', qty: 2}
    ];
    const actionToDispatch = updatePortfolio(mockPortfolio);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.updatePortfolio(mockPortfolio);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});

describe('mapStateToProps', () => {
  it('should return portfolio obj', () => {
    const mockState = {
      portfolio: [{name: 'Bitcoin', qty: 1}, {name: 'Litecoin', qty: 3}],
      trackedCoins: ['BTC','LTC'],
    };

    const expected = {
      portfolio: [{name: 'Bitcoin', qty: 1}, {name: 'Litecoin', qty: 3}]
    };

    const mappedState = mapStateToProps(mockState);
    expect(mappedState).toEqual(expected);
  });
});
