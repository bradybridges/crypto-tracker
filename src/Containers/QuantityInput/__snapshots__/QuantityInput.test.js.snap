// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuantityInput should match the snapshot 1`] = `
<div
  className="coin-input-div"
>
  <label
    className="coin-label"
  >
    Bitcoin
  </label>
  <input
    className="qty-input"
    onChange={[Function]}
    placeholder="Quantity: (5)"
    type="number"
    value=""
  />
</div>
`;
